### ISSUE TITLE
Significant Speedup of Kanji Search

### ISSUE THREAD

**Issue by Garethp** [melink14/rikaikun#145]
So, I noticed Kanji Searches in RikaiRebuilt was actually fairly slow. Makes sense, every time it needs to find information for the Kanji it searches the 12,500 line string line by line looking for the Kanji. After taking a look at what it did, it was basically just returning whatever line starts with the kanji given. So I managed to replace that function with something much more readable and faster. Maybe you want to test it and see how you like it.

My Solution:
```js
if (!this.kanjiData) {
    this.kanjiData = await FileReader.read('../resources/kanji.dat')
        .then(response => response.text())
        .then(data => data.split('\n').filter(line => line.length !== 0))
        .then(array => array.reduce((obj, current) => {
            obj[current.charAt(0)] = current;
            return obj;
        }, {}));
}

kde = this.kanjiData[character];
if (!kde) return null;

```

So, pretty simply, rather than just read the entire file in to string and then call `this.find` on that string I load the string up, split it in to an array and then reduce the array in to an obj where the key is the kanji character. Bam, instant indexing for searching. Seems much much faster for me. I'll probably take it a bit further at some point and pre-prepare the actual result (because the `kanjiSearch` function is ugly as hell) and maybe even write the entire thing as a JSON so there's no need to process on the client at all. But for now, this should be useful for you as well

**Comment by melink14** [comment]
I missed this when it was first posted but thanks for the idea.

I agree that loading it into memory first should be much faster and not too demanding of modern computers. Not sure if I would take this intermediate step or move forward with something like indexDB to prepare for Manifest v3.

That being said, were you seeing noticeble slow downs? In rikaikun it seems instaneous so I never prioritized the speed.

**Comment by Garethp** [comment]
It's been a fairly long time, so I don't remember how slow I thought it was to begin with, but I remember that after I changed it to this way the *difference* was very noticeable, at least in my application.

Moving to IndexDB is probably a decent idea. I use IndexDB for regular word lookup, but I suppose I didn't consider needing to do it for Kanji. 

I think one of the other main reasons I changed it to this way was more for readability. I found the original method hard to follow and understand, whereas this made much more sense to me

**Comment by melink14** [comment]
Thanks for the reply. That makes sense.

Most of the still original code hasn't aged well so you're definitely right on the readability part.

I'll probably leave this open as a reminder that the Kanji logic needs work.
